stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4182 @@
+        Ok(self
+            .tenure_blocks
+            .take()
+            .map(|blocks| blocks.into_iter().rev().collect()))
+    }
+
+    /// Produce the next HTTP request that, when successfully executed, will fetch the data needed
+    /// to advance this state machine.
+    /// Not all states require an HTTP request for advanceement.
+    ///
+    /// Returns Ok(Some(request)) if a request is needed
+    /// Returns Ok(None) if a request is not needed (i.e. we're waiting for some other machine's
+    /// state)
+    /// Returns Err(()) if we're done.
+    pub fn make_next_download_request(
+        &self,
+        peerhost: PeerHost,
+    ) -> Result<Option<StacksHttpRequest>, ()> {
+        let request = match self.state {
+            NakamotoTenureDownloadState::GetTenureStartBlock(start_block_id) => {
+                test_debug!("Request tenure-start block {}", &start_block_id);
+                StacksHttpRequest::new_get_nakamoto_block(peerhost, start_block_id.clone())
+            }
+            NakamotoTenureDownloadState::WaitForTenureEndBlock(_block_id) => {
+                // we're waiting for some other downloader's block-fetch to complete
+                test_debug!("Waiting for tenure-end block {}", &_block_id);
+                return Ok(None);
+            }
+            NakamotoTenureDownloadState::GetTenureEndBlock(end_block_id) => {
+                test_debug!("Request tenure-end block {}", &end_block_id);
+                StacksHttpRequest::new_get_nakamoto_block(peerhost, end_block_id.clone())
+            }
+            NakamotoTenureDownloadState::GetTenureBlocks(end_block_id) => {
+                test_debug!("Downloading tenure ending at {}", &end_block_id);
+                StacksHttpRequest::new_get_nakamoto_tenure(peerhost, end_block_id.clone(), None)
+            }
+            NakamotoTenureDownloadState::Done => {
+                // nothing more to do
+                return Err(());
+            }
+        };
+        Ok(Some(request))
+    }
+
+    /// Begin the next download request for this state machine.  The request will be sent to the
+    /// data URL corresponding to self.naddr.
+    /// Returns Ok(Some(true)) if we sent the request, or there's already an in-flight request
+    /// Returns Ok(Some(false)) if not (e.g. neighbor is known to be dead or broken)
+    /// Returns Ok(None) if there is already an in-flight request to this peer.  The caller should
+    /// try again.
+    /// Returns Err(..) if self.naddr is known to be a dead or broken peer, or if we were unable to
+    /// resolve its data URL to a socket address.
+    pub fn send_next_download_request(
+        &mut self,
+        network: &mut PeerNetwork,
+        neighbor_rpc: &mut NeighborRPC,
+    ) -> Result<Option<bool>, NetError> {
+        if neighbor_rpc.has_inflight(&self.naddr) {
+            test_debug!("Peer {} has an inflight request", &self.naddr);
+            return Ok(Some(true));
```

Comment:
Outdated

Reply:
Hmm -- but the update removed the distinction between `Ok(None)` and `Ok(Some(true))` which seems like its actually important. If there's an in flight request to the peer, but its not for this downloader, doesn't that mean that this downloader _isn't_ currently doing work?

Reply:
The system determines whether or not the downloader is doing work via the `.idle` member, not this function.  The caller is really interested in the `Err(..)` and `Ok(false)` cases, since these indicate that the downloader's remote peer has broken its connection and/or malfunctioned, or the downloader has reached its halting state.  `Ok(true)` means "try again"